### ISSUE TITLE
Add support for broadcasting learning rate scheduler in PyTorch

### ISSUE THREAD

**Issue by tgaddair** [horovod/horovod#1281]
**Is your feature request related to a problem? Please describe.**
Some users wish to broadcast the learning rate scheduler similar to how we support `broadcast_optimizer_state`, but there is no out of the box solution for this.

**Describe the solution you'd like**
Ideally, we would either have a new function `broadcast_lr_scheduler_state` that does just this for a learning rate scheduler, or we could bring back support for broadcasting arbitrary Python objects.

**Describe alternatives you've considered**
Having a function to broadcast any serializable object can be useful and general, but sometimes unclear to the user, and slow if implemented naively.  Users can workaround this issue by rolling their own byte tensor serialization and broadcast, but that's tedious.

**Additional context**
We previously had support for [broadcast_object](https://github.com/horovod/horovod/pull/371/files#diff-b134e260de34b7f653af3c43e362f305L124), but this function was removed when we changed the implementation of `broadcast_optimizer_state`, as pickling was deemed too inefficient for optimizer state.

**Comment by hdubey** [comment]
@tgaddair Is there any way to use the Learning rate schedular for pytorch with horovod? If we choose to not broadcast it, will it cause poor learning?

**Comment by tgaddair** [comment]
Hey @hdubey, it should work fine with Horovod.  The only tricky part is if you need to checkpoint your learning rate scheduler state so you can pause and resume training.  Currently, you'd need to write/read it in on every worker separately, but ideally it should work the same as our other broadcast calls, where you can write/read on worker-0 then broadcast the state to the other workers.

**Comment by kaituoxu** [comment]
any update?

**Comment by tgaddair** [comment]
@kaituoxu @hdubey can you try using the [broadcast_object](https://github.com/horovod/horovod/blob/master/horovod/torch/__init__.py#L598) function in Horovod 0.19.0 with learning rate schedules?

**Comment by kaituoxu** [comment]
Cool~

**Comment by LearnedVector** [comment]
@tgaddair  @kaituoxu any update if this works?

**Comment by tgaddair** [comment]
Hey @LearnedVector, the `broadcast_object` function is available in Horovod v0.19.0.  Can you give it a shot and let me know if it works for you?  I haven't run into any issues in my tests.

**Comment by bakalaugh** [comment]
Even though it's super earlier issue, but as saw it is still opening, wanna give a conclusion here :).

1) For the **scheduler unrelated to the loss value but only to the number of epoch or step**: there is no need to broadcast the `state_dict` of scheduler, as the epoch or step number will be automatically used to adjust the `lr` in each process' script, which is always consistent.

2) For the **scheduler related to the loss value**, such as `ReduceLROnPlateau`: one can do allreduce operation over `loss` value across all processes to obtain average `avg_loss`, and then use `scheduler.step(avg_loss)` as usual. Then every scheduler will perform consistently. 

***

The allreduce operation can refer to the tutorial https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_mnist.py#L144/.

Here is just a dummy script i used for testing. With allreduce, the best value within scheduler `state_dict` is consistent and vice versa.
```
  def _test_scheduler():
      class test_model(nn.Module):
          def __init__(self):
              super().__init__()
              self.fc = nn.Linear(3, 3)

          def forward(self, x):
              return self.fc(x)
          
      # alter: with allreduce
      model = test_model()
      optim = torch.optim.Adam(model.parameters())
      sched = torch.optim.lr_scheduler.ReduceLROnPlateau(optim)

      x = torch.tensor([[0.1, 0.2, 0.3], [-0.1, -0.2, -0.3]])
      
      for i in range(30):
          out = model(x)
          loss = out.sum()
          avg_loss = hvd.allreduce(loss, name='test')
          
          if i == 0:
              sched.step(avg_loss - 1)
          sched.step(avg_loss)

      print(f'optim_state (with): {optim.param_groups[0].get("lr")}')
      print(f'sched_state (with): {sched.state_dict().get("best")}')
      
      # alter: without allreduce
      model = test_model()
      optim = torch.optim.Adam(model.parameters())
      sched = torch.optim.lr_scheduler.ReduceLROnPlateau(optim)

      x = torch.tensor([[0.1, 0.2, 0.3], [-0.1, -0.2, -0.3]])
      
      for i in range(30):
          out = model(x)
          loss = out.sum()
          avg_loss = loss
          
          if i == 0:
              sched.step(avg_loss - 1)
          sched.step(avg_loss)

      print(f'optim  (without): {optim.param_groups[0].get("lr")}')
      print(f'sched best (without): {sched.state_dict().get("best")}')
```
